compiler/rustc_parse_format/src/lib.rs

```diff
@@ -483,34 +488,32 @@ impl<'a> Parser<'a> {
                 None
             }
         } else {
-            let description = format!("expected `{c:?}` but string was terminated");
+            let description = format!("expected `'}}'` but string was terminated");
             // point at closing `"`
             let pos = self.input.len() - if self.append_newline { 1 } else { 0 };
             let pos = self.to_span_index(pos);
-            if c == '}' {
-                let label = format!("expected `{c:?}`");
-                let (note, secondary_label) = if c == '}' {
-                    (
-                        Some(
-                            "if you intended to print `{`, you can escape it using `{{`".to_owned(),
-                        ),
-                        self.last_opening_brace
-                            .map(|sp| ("because of this opening brace".to_owned(), sp)),
-                    )
-                } else {
-                    (None, None)
-                };
-                self.errors.push(ParseError {
-                    description,
-                    note,
-                    label,
-                    span: pos.to(pos),
-                    secondary_label,
-                    should_be_replaced_with_positional_argument: false,
-                });
+            let label = format!("expected `'}}'`");
+            let (note, secondary_label) = if arg.format.fill == Some('}') {
+                (
+                    Some("the character `}` is interpreted as a fill character because of the `:` that precedes it".to_owned()),
+                    arg.format.fill_span.map(|sp| ("this is not interpreted as a formatting closing brace".to_owned(), sp)),
+                )
             } else {
-                self.err(description, format!("expected `{c:?}`"), pos.to(pos));
-            }
+                (
+                    Some("if you intended to print `{`, you can escape it using `{{`".to_owned()),
+                    self.last_opening_brace
+                        .map(|sp| ("because of this opening brace".to_owned(), sp)),
+                )
+            };
```

Comment:
It's a little unfortunate to have this code basically duplicated from above, but I suspect the fix might be worse. I guess it might be possible to lift this into a closure and call it in both cases, but I think that would be harder to follow. What do you think?